### PR TITLE
Support MS SQL Server Data Source without DSN using odbc package

### PR DESCRIPTION
**Pull request by hidekoji**
# Description

Support MS SQL Server Data Source without DSN using odbc package

# Checklist
Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [x] Pass devtools::check()
- [ ] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/system.R [hunk 1]
```diff
@@ -490,7 +490,7 @@ clearAmazonAthenaConnection <- function(driver = "", region = "", authentication
 #' If not, new connection is created and returned.
 #' @export
 getDBConnection <- function(type, host = NULL, port = "", databaseName = "", username = "", password = "", catalog = "", schema = "", dsn="", additionalParams = "",
-                            collection = "", isSSL = FALSE, authSource = NULL, cluster = NULL, timeout = NULL, connectionString = NULL) {
+                            collection = "", isSSL = FALSE, authSource = NULL, cluster = NULL, timeout = NULL, connectionString = NULL, driver = "") {
```
Comment (hideaki) [inline R/system.R]: Why is the default for driver set to "" rather than NA?

File: R/system.R [hunk 2]
```diff
@@ -705,6 +705,50 @@ getDBConnection <- function(type, host = NULL, port = "", databaseName = "", use
         connection_pool[[key]] <- conn
       }
     }
+  } else if (type == "mssqlserver") {# The type sqlserver is already used for RODBC based one and "mssqlserver" is passed from Exploratory Desktop.
+
+    # if platform is Linux use predefined one
```
Comment (hideaki) [inline R/system.R]: Please add comment on the reason of this.

File: R/system.R [hunk 3]
```diff
@@ -991,19 +1039,30 @@ queryAmazonAthena <- function(driver = "", region = "", authenticationType = "IA
 #' @param port - Database port number
 #' @param as.is - Flag to tell if you honor data types from ODBC
 #'
-queryODBC <- function(dsn,username, password, additionalParams, numOfRows = 0, query, stringsAsFactors = FALSE, host="", port="", as.is = TRUE, ...){
-  if(!requireNamespace("RODBC")){stop("package RODBC must be installed.")}
-
-  conn <- getDBConnection("odbc", host, port, NULL, username, password, dsn = dsn, additionalParams = additionalParams)
+queryODBC <- function(dsn="", username, password, additionalParams="", numOfRows = 0, query, stringsAsFactors = FALSE, host="", port="", as.is = TRUE, databaseName="", driver = "", type = "", ...){
```
Comment (hideaki) [inline R/system.R]: Please explain when type can be other than "odbc" here in the comment.

File: R/system.R [hunk 4]
```diff
@@ -1017,9 +1076,12 @@ queryODBC <- function(dsn,username, password, additionalParams, numOfRows = 0, q
   }, error = function(err) {
     # for some cases like conn not being an open connection, sqlQuery still throws error. handle it here.
     # clear connection in pool so that new connection will be used for the next try
-    clearDBConnection("odbc", NULL, NULL, NULL, username, dsn = dsn, additionalParams = additionalParams)
+    clearDBConnection(type, NULL, NULL, NULL, username, dsn = dsn, additionalParams = additionalParams)
     stop(err)
   })
+  if(type == "mssqlserver") {
+    DBI::dbClearResult(resultSet)
```
Comment (hideaki) [inline R/system.R]: Please explain why this is necessary only for mssqlserver case, if that is the case.

File: R/system.R [hunk 5]
```diff
@@ -705,6 +705,50 @@ getDBConnection <- function(type, host = NULL, port = "", databaseName = "", use
...
       }
     }
+  } else if (type == "mssqlserver") {# The type sqlserver is already used for RODBC based one and "mssqlserver" is passed from Exploratory Desktop.
+
+    # if platform is Linux use predefined one
+    if(Sys.info()["sysname"]=="Linux"){
+      driver <-  "ODBC Driver 17 for SQL Server";
+    }
+    if(!requireNamespace("DBI")){stop("package DBI must be installed.")}
+    if(!requireNamespace("odbc")){stop("package RPostgres must be installed.")}
```
Comment (hideaki) [inline R/system.R]: Please add comment on why we are using odbc instead of RODBC here.

**Comment by hidekoji** [comment]
Tested connection failure and fallback

```r
Error in connection_sql_tables(conn@ptr, catalog_name = catalog_name,  :
  nanodbc/nanodbc.cpp:4357: 00000: [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: Error code 0x274C  [Microsoft][ODBC Driver 17 for SQL Server]Communication link failure 
```

And I can get a valid connection after that.

File: R/system.R (hunk 1 above)
Comment (hidekoji) [reply R/system.R]: Changed it to NULL

File: R/system.R (hunk 2 above)
Comment (hidekoji) [reply R/system.R]: Added comment.

File: R/system.R (hunk 3 above)
Comment (hidekoji) [reply R/system.R]: Added comment.

File: R/system.R (hunk 4 above)
Comment (hidekoji) [reply R/system.R]: Added comment.

File: R/system.R (hunk 5 above)
Comment (hidekoji) [reply R/system.R]: Added comment.

**Pull request merged**